rfcs/0021-pr-testing.md

```diff
@@ -0,0 +1,85 @@
+
+# Detailed design
+[design]: #detailed-design
+
+The general idea is to have a service which receives webhooks from github,
+then requires some acknowledgement of a maintainer to build/test a PR and
+subsequently schedules a jobset on Hydra.
+
+## Details of the reference implementation at https://github.com/mayflower/nixborg
+(to be moved to the NixOS github organisation if accepted)
+
+The current implementation consists of three seperate services:
+ - nixborg: flask app to receive github hooks and fill a queue (currently redis, probably switching to rabbitmq in the future)
+ - nixborg-workers: workers consuming the queue, doing the git handling, commenting on github, HTTP requesting to nixbot-receiver
+ - nixbot-receiver: stdlib-dependency-only python script to be run on Hydra that calls out to hydra-update-jobset (PR: https://github.com/NixOS/hydra/pull/516)
+
+1. the main service `nixborg` listens for `issue_comment` hooks containing `@nixborg build`
+2. `nixborg` passes the hook on to the `nixborg-workers`
+3. a `nixborg-worker` rebases the PR on the base branch of the PR (mainly master, staging or release-XX.XX)
+4. the rebased branch gets pushed to a `pr-XXXXX` branch
+5. a `nixborg-worker` HTTP calls out to the `nixbot-receiver` on Hydra with all necessary information for a jobset; this message is HMAC authenticated
+6. `nixbot-receiver` passes on the information to create/update a jobset `pr-XXXXX` with the newly implemented `hydra-update-jobset` CLI utility
+
+Some small remarks on the architecture:
+ - the rebasing is done in order to:
+   - make sure Hydra builds a commit that has been reviewed by a maintainer
+   - ensure an older PR is tested based on the current state of the base branch
+ - `nixbot-receiver` is implemented as a stdlib-dependency-only python script not doing any logic to have a stable interface to Hydra that doesn't need to be updated
+    often, when iterating on nixborg and to not have a large footprint (cpu, memory, disk space)
+
+# Alternatives
+[alternatives]: #alternatives
+
+ - The Hydra jobset management could be achieved by using declarative jobsets, this does not work out currently as jobsets cannot be compared across projects
+ - [bors](https://bors.tech/) is the bot the Rust developers use to automate their testing with buildbot. It would have to be adapted to call out to Hydra
+   and currently AFAIK is not that easy to run yourself, when not using heroku. A lot of nixborg is inspired by bors.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+ - Which "release.nix" should be used to test:
+   - We probably want to test aarch64 and x86_64-darwin, not only x86_64-linux, as only a little number of maintainers/pull requesters have access to these architectures.
+   - We probably want to also test NixOS module tests, especially for PRs touching these
+ - Can Hydra handle the additional workload by this, we probably have to start with a small number of PRs being tested to evaluate this.
```

Comment:
That, and maybe perform initial tests only `x86_64-linux`? How the Hydra capacity (or maybe scaling costs, if cloud builds are used regularly — I do not know) compare across targets?